README.md
## Page Monitor

> capture webpage and diff the dom change with [phantomjs](http://phantomjs.org/)

## Effects

### Element Add

![element add](./demo/1409037825746-1409037838093.png)

### Element Removed

![element removed](./demo/1409037838093-1409037882033.png)

### Text Changed

![element removed](./demo/1409037882033-1409037916727.png)

### Style Changed

![element removed](./demo/1409038130483-1409038137417.png)

## Make UI to show diffs of history

> https://github.com/fouber/pmui

![ui](./demo/ui.png?v=3)

## Usage

> First of all, install [phantomjs](http://phantomjs.org/download.html), page-monitor relys on [phantomjs](http://phantomjs.org/) to render webpage and genenrate screenshot. DO NOT use ``npm`` to install phantomjs on winidows, it can't be launched by ``spawn``.

```javascript
var Monitor = require('page-monitor');

var url = 'http://www.google.com';
var monitor = new Monitor(url);
monitor.capture(function(code){
    console.log(monitor.log); // from phantom
    console.log('done, exit [' + code + ']');
});
```

## API

### Monitor

```javascript
var monitor = new Monitor(url [, options]);
```

see the default options here: [https://github.com/fouber/page-monitor/blob/master/index.js](https://github.com/fouber/page-monitor/blob/master/index.js#L57-L179) , you can override any option for your monitoring.

### monitor.capture(callback [, noDiff]);

caputure webpage and save screenshot, then diff with last save.

```javascript
var monitor = new Monitor(url, options);
monitor.on('debug', function (data) {
    console.log('[DEBUG] ' + data);
});
monitor.on('error', function (data) {
    console.error('[ERROR] ' + data);
});
monitor.capture(function(code){
    console.log(monitor.log.info); // diff result
    console.log('[DONE] exit [' + code + ']');
});
```

### monitor.diff(left, right, callback);

diff change between left(date.getTime()) and right(date.getTime()).

```javascript
var monitor = new Monitor(url, options);
monitor.on('debug', function (data) {
    console.log('[DEBUG] ' + data);
});
monitor.on('error', function (data) {
    console.error('[ERROR] ' + data);
});
monitor.diff(1408947323420, 1408947556898, function(code){
    console.log(monitor.log.info); // diff result
    console.log('[DONE] exit [' + code + ']');
});
```

### events

```javascript
var monitor = new Monitor(url);
monitor.on('debug', function (data) {
    console.log('[DEBUG] ' + data);
});
monitor.on('error', function (data) {
    console.error('[ERROR] ' + data);
});
```

* ``debug``: debug from phantom
* ``notice``: console from webpage
* ``info``: info from phantom
* ``warning``: error from webpage
* ``error``: error from phantom


CHANGELOG.md
## Thu Sep 25 2014

> Because of screenshot is png format, it will occupy so much space, so I change it into ``jpeg`` and set quality as ``80`` by default.

* [Mod] save screenshot as ``jpeg`` by default, and quality is ``80``
* [Add] add options to set screenshot format, quality.

If you want to save screenshot as png format, you can:


```javascript
var Monitor = require('monitor');
var m = new Monitor('http://www.example.com', {
    render: {
        format: 'png'
    }
});
```

util.js
var toString = Object.prototype.toString;
var _ = module.exports = {};

/**
 * is
 * @param {*} source
 * @param {string} type
 * @returns {boolean}
 */
_.is = function(source, type){
    return toString.call(source) === '[object ' + type + ']';
};

/**
 * walk object and merge
 * @param {Object} obj
 * @param {Function} callback
 * @param {Object} merge
 */
_.map = function(obj, callback, merge){
    var index = 0;
    for(var key in obj){
        if(obj.hasOwnProperty(key)){
            if(merge){
                callback[key] = obj[key];
            } else if(callback(key, obj[key], index++)) {
                break;
            }
        }
    }
};

/**
 * merge target into source
 * @param {*} source
 * @param {*} target
 * @returns {*}
 */
_.merge = function(source, target){
    if(_.is(source, 'Object') && _.is(target, 'Object')){
        _.map(target, function(key, value){
            source[key] = _.merge(source[key], value);
        });
    } else {
        source = target;
    }
    return source;
};

/**
 * escape regexp chars
 * @param {string} str
 * @returns {string}
 */
_.escapeReg = function(str){
    return str.replace(/[\.\\\+\*\?\[\^\]\$\(\){}=!<>\|:\/]/g, '\\$&');
};

/**
 * pad a numeric string to two digits
 * @param {string} str
 * @returns {string}
 */
_.pad = function(str){
    return ('0' + str).substr(-2);
};

/**
 * convert millisecond into string like `2014-09-12 14:23:03`
 * @param {Number} num
 * @returns {string}
 */
_.getTimeString = function(num){
    var d;
    if(_.is(num, 'Date')){
        d = num;
    } else {
        d = new Date();
        d.setTime(num);
    }
    var day = [
        d.getFullYear(),
        _.pad(d.getMonth() + 1),
        _.pad(d.getDate())
    ].join('-');
    var time = [
        _.pad(d.getHours()),
        _.pad(d.getMinutes()),
        _.pad(d.getSeconds())
    ].join(':');
    return day + ' ' + time;
};

/**
 * generate UUID
 * @returns {string}
 */
_.unique = function(){
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        var r = Math.random()*16 | 0;
        return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
    });
};

/**
 * run mode, capture or diff or both
 * @type {{CAPTURE: number, DIFF: number}}
 */
_.mode = {
    CAPTURE: 1, // capture mode
    DIFF   : 2  // diff mode
};

/**
 * log type
 * @type {{DEBUG: string, WARNING: string, INFO: string, ERROR: string, NOTICE: string}}
 */
_.log = {
    DEBUG: '<[debug]>',
    WARNING: '<[warning]>',
    INFO: '<[info]>',
    ERROR: '<[error]>',
    NOTICE: '<[notice]>'
};

phantomjs/diff.js
/**
 * equal
 * @param {Object} left
 * @param {Object} right
 * @returns {boolean}
 */
function equal(left, right){
    var type = typeof left;
    if(type === typeof right){
        switch(type){
            case 'object':
                var lKeys = Object.keys(left);
                var rKeys = Object.keys(right);
                if(lKeys.length === rKeys.length){
                    for(var i = 0; i < lKeys.length; i++){
                        var key = lKeys[i];
                        if(!right.hasOwnProperty(key) || (left[key] !== right[key])){
                            return false;
                        }
                    }
                    return true;
                } else {
                    return false;
                }
                break;
            default:
                return left === right;
        }
    } else {
        return false;
    }
}

/**
 * match
 * @param {Object} left
 * @param {Object} right
 * @returns {boolean}
 */
function isMatch(left, right){
    return (left.name === right.name) && equal(left.attr, right.attr);
}

/**
 * common logic of `LCSHeadFirst’ and `LCSTailFirst‘
 * @param {Object} old
 * @param {Object} cur
 * @param {Function} match
 * @param {Number} x
 * @param {Array} lastLine
 * @param {Array} currLine
 */
function LCSProc(old, cur, match, x, lastLine, currLine){
    if(match(old, cur)){
        var sequence = (lastLine[x-1] || []).slice(0);
        sequence.push({ l: old, r: cur });
        currLine[x] = sequence;
    } else {
        var lSeq = currLine[x-1];
        var tSeq = lastLine[x];
        if(lSeq && tSeq){
            if(lSeq.length < tSeq.length){
                currLine[x] = tSeq.slice(0);
            } else {
                currLine[x] = lSeq.slice(0);
            }
        } else if(lSeq) {
            currLine[x] = lSeq.slice(0);
        } else if(tSeq) {
            currLine[x] = tSeq.slice(0);
        }
    }
}

/**
 * Longest common subsequence (obverse)
 * @param {Array} left
 * @param {Array} right
 * @param {Function} match
 * @returns {Array}
 */
function LCSHeadFirst(left, right, match){
    var lastLine = [];
    var currLine = [];
    var y = left.length;
    var len = right.length;
    while(y--){
        var old = left[y];
        var i = len;
        while(i--){
            var cur = right[i];
            var x = len -  i - 1;
            LCSProc(old, cur, match, x, lastLine, currLine);
        }
        lastLine = currLine;
        currLine = [];
    }
    return (lastLine.pop() || []);
}

/**
 * Longest common subsequence (reverse)
 * @param {Array} left
 * @param {Array} right
 * @param {Function} match
 * @returns {Array}
 */
function LCSTailFirst(left, right, match){
    var lastLine = [];
    var currLine = [];
    left.forEach(function(old){
        right.forEach(function(cur, x){
            LCSProc(old, cur, match, x, lastLine, currLine);
        });
        lastLine = currLine;
        currLine = [];
    });
    return (lastLine.pop() || []);
}

/**
 * diff change
 * @param {Object} left
 * @param {Object} right
 * @param {Object} opt
 * @returns {Array}
 */
var diff = function(left, right, opt){
    var ret = [];
    var change = {
        type: 0,
        node: right
    };
    if(left.style !== right.style){
        change.type |= opt.changeType.STYLE;
    }
    var LCS = opt.priority === 'head' ? LCSHeadFirst : LCSTailFirst;
    LCS(left.child, right.child, isMatch).forEach(function(node){
        var old = node.l;
        var cur = node.r;
        cur.matched = old.matched = true;
        if(cur.name === '#'){
            if(old.text !== cur.text){
                // match node, but contents are different.
                change.type |= opt.changeType.TEXT;
            }
        } else {
            // recursive
            ret = ret.concat(diff(old, cur, opt));
        }
    });
    right.child.forEach(function(node){
        if(!node.matched){
            if(node.name === '#'){
                // add text, but count as text change
                change.type |= opt.changeType.TEXT;
            } else {
                // add element
                ret.push({
                    type: opt.changeType.ADD,
                    node: node
                });
            }
        }
    });
    left.child.forEach(function(node){
        if(!node.matched){
            if(node.name === '#'){
                // remove text, but count as text change
                change.type |= opt.changeType.TEXT;
            } else {
                // removed element
                ret.push({
                    type: opt.changeType.REMOVE,
                    node: node
                });
            }
        }
    });
    if(change.type){
        ret.push(change);
    }
    return ret;
};

module.exports = diff;

phantomjs/walk.js
module.exports = function(TOKEN, data){

    /*
     * JavaScript MD5 1.0.1
     * https://github.com/blueimp/JavaScript-MD5
     *
     * Copyright 2011, Sebastian Tschan
     * https://blueimp.net
     *
     * Licensed under the MIT license:
     * http://www.opensource.org/licenses/MIT
     *
     * Based on
     * A JavaScript implementation of the RSA Data Security, Inc. MD5 Message
     * Digest Algorithm, as defined in RFC 1321.
     * Version 2.2 Copyright (C) Paul Johnston 1999 - 2009
     * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
     * Distributed under the BSD License
     * See http://pajhome.org.uk/crypt/md5 for more info.
     */

    var md5 = function(){"use strict";function n(n,r){var t=(65535&n)+(65535&r),u=(n>>16)+(r>>16)+(t>>16);return u<<16|65535&t}function r(n,r){return n<<r|n>>>32-r}function t(t,u,e,o,c,f){return n(r(n(n(u,t),n(o,f)),c),e)}function u(n,r,u,e,o,c,f){return t(r&u|~r&e,n,r,o,c,f)}function e(n,r,u,e,o,c,f){return t(r&e|u&~e,n,r,o,c,f)}function o(n,r,u,e,o,c,f){return t(r^u^e,n,r,o,c,f)}function c(n,r,u,e,o,c,f){return t(u^(r|~e),n,r,o,c,f)}function f(r,t){r[t>>5]|=128<<t%32,r[(t+64>>>9<<4)+14]=t;var f,i,a,h,g,l=1732584193,v=-271733879,d=-1732584194,C=271733878;for(f=0;f<r.length;f+=16)i=l,a=v,h=d,g=C,l=u(l,v,d,C,r[f],7,-680876936),C=u(C,l,v,d,r[f+1],12,-389564586),d=u(d,C,l,v,r[f+2],17,606105819),v=u(v,d,C,l,r[f+3],22,-1044525330),l=u(l,v,d,C,r[f+4],7,-176418897),C=u(C,l,v,d,r[f+5],12,1200080426),d=u(d,C,l,v,r[f+6],17,-1473231341),v=u(v,d,C,l,r[f+7],22,-45705983),l=u(l,v,d,C,r[f+8],7,1770035416),C=u(C,l,v,d,r[f+9],12,-1958414417),d=u(d,C,l,v,r[f+10],17,-42063),v=u(v,d,C,l,r[f+11],22,-1990404162),l=u(l,v,d,C,r[f+12],7,1804603682),C=u(C,l,v,d,r[f+13],12,-40341101),d=u(d,C,l,v,r[f+14],17,-1502002290),v=u(v,d,C,l,r[f+15],22,1236535329),l=e(l,v,d,C,r[f+1],5,-165796510),C=e(C,l,v,d,r[f+6],9,-1069501632),d=e(d,C,l,v,r[f+11],14,643717713),v=e(v,d,C,l,r[f],20,-373897302),l=e(l,v,d,C,r[f+5],5,-701558691),C=e(C,l,v,d,r[f+10],9,38016083),d=e(d,C,l,v,r[f+15],14,-660478335),v=e(v,d,C,l,r[f+4],20,-405537848),l=e(l,v,d,C,r[f+9],5,568446438),C=e(C,l,v,d,r[f+14],9,-1019803690),d=e(d,C,l,v,r[f+3],14,-187363961),v=e(v,d,C,l,r[f+8],20,1163531501),l=e(l,v,d,C,r[f+13],5,-1444681467),C=e(C,l,v,d,r[f+2],9,-51403784),d=e(d,C,l,v,r[f+7],14,1735328473),v=e(v,d,C,l,r[f+12],20,-1926607734),l=o(l,v,d,C,r[f+5],4,-378558),C=o(C,l,v,d,r[f+8],11,-2022574463),d=o(d,C,l,v,r[f+11],16,1839030562),v=o(v,d,C,l,r[f+14],23,-35309556),l=o(l,v,d,C,r[f+1],4,-1530992060),C=o(C,l,v,d,r[f+4],11,1272893353),d=o(d,C,l,v,r[f+7],16,-155497632),v=o(v,d,C,l,r[f+10],23,-1094730640),l=o(l,v,d,C,r[f+13],4,681279174),C=o(C,l,v,d,r[f],11,-358537222),d=o(d,C,l,v,r[f+3],16,-722521979),v=o(v,d,C,l,r[f+6],23,76029189),l=o(l,v,d,C,r[f+9],4,-640364487),C=o(C,l,v,d,r[f+12],11,-421815835),d=o(d,C,l,v,r[f+15],16,530742520),v=o(v,d,C,l,r[f+2],23,-995338651),l=c(l,v,d,C,r[f],6,-198630844),C=c(C,l,v,d,r[f+7],10,1126891415),d=c(d,C,l,v,r[f+14],15,-1416354905),v=c(v,d,C,l,r[f+5],21,-57434055),l=c(l,v,d,C,r[f+12],6,1700485571),C=c(C,l,v,d,r[f+3],10,-1894986606),d=c(d,C,l,v,r[f+10],15,-1051523),v=c(v,d,C,l,r[f+1],21,-2054922799),l=c(l,v,d,C,r[f+8],6,1873313359),C=c(C,l,v,d,r[f+15],10,-30611744),d=c(d,C,l,v,r[f+6],15,-1560198380),v=c(v,d,C,l,r[f+13],21,1309151649),l=c(l,v,d,C,r[f+4],6,-145523070),C=c(C,l,v,d,r[f+11],10,-1120210379),d=c(d,C,l,v,r[f+2],15,718787259),v=c(v,d,C,l,r[f+9],21,-343485551),l=n(l,i),v=n(v,a),d=n(d,h),C=n(C,g);return[l,v,d,C]}function i(n){var r,t="";for(r=0;r<32*n.length;r+=8)t+=String.fromCharCode(n[r>>5]>>>r%32&255);return t}function a(n){var r,t=[];for(t[(n.length>>2)-1]=void 0,r=0;r<t.length;r+=1)t[r]=0;for(r=0;r<8*n.length;r+=8)t[r>>5]|=(255&n.charCodeAt(r/8))<<r%32;return t}function h(n){return i(f(a(n),8*n.length))}function g(n,r){var t,u,e=a(n),o=[],c=[];for(o[15]=c[15]=void 0,e.length>16&&(e=f(e,8*n.length)),t=0;16>t;t+=1)o[t]=909522486^e[t],c[t]=1549556828^e[t];return u=f(o.concat(a(r)),512+8*r.length),i(f(c.concat(u),640))}function l(n){var r,t,u="0123456789abcdef",e="";for(t=0;t<n.length;t+=1)r=n.charCodeAt(t),e+=u.charAt(r>>>4&15)+u.charAt(15&r);return e}function v(n){return unescape(encodeURIComponent(n))}function d(n){return h(v(n))}function C(n){return l(d(n))}function A(n,r){return g(v(n),v(r))}function m(n,r){return l(A(n,r))}function s(n,r,t){return r?t?A(r,n):m(r,n):t?d(n):C(n)}return s}();

    /**
     * combo selectors
     * @param {string} selectors
     * @returns {string}
     */
    function normalizeSelectors(selectors){
        if(Object.prototype.toString.call(selectors) === '[object Array]'){
            return selectors.join(',');
        } else {
            return String(selectors || '');
        }
    }

    // walk settings
    var INVISIBLE_ELEMENT = data.invisibleElements;
    var IGNORE_CHILDREN_ELEMENT = data.ignoreChildrenElements;
    var STYLE_FILTERS = data.styleFilters;
    var ATTR_FILTERS = data.attributeFilters;
    var INCLUDE_SELECTORS = normalizeSelectors(data.includeSelectors);
    var EXCLUDE_SELECTORS = normalizeSelectors(data.excludeSelectors);
    var IGNORE_CHILDREN_SELECTORS = normalizeSelectors(data.ignoreChildrenSelectors);
    var IGNORE_TEXT_SELECTORS = normalizeSelectors(data.ignoreTextSelectors);
    var ROOT = data.root || 'body';

    // reg
    var invisibleElementReg = new RegExp('^(' + INVISIBLE_ELEMENT.join('|') + ')$', 'i');
    var ignoreChildrenElementReg = new RegExp('^(' + IGNORE_CHILDREN_ELEMENT.join('|') + ')$', 'i');

    /**
     * invisible
     * @param {HTMLElement} elem
     * @returns {boolean}
     */
    function isInvisible(elem){
        var tagName = elem.tagName.toLowerCase();
        invisibleElementReg.lastIndex = 0;
        return (invisibleElementReg.test(tagName) || (tagName === 'input' && elem.type === 'hidden'));
    }

    /**
     * ignore child
     * @param {HTMLElement} elem
     * @returns {boolean}
     */
    function igonreChildren(elem){
        ignoreChildrenElementReg.lastIndex = 0;
        return ignoreChildrenElementReg.test(elem.tagName) ||
              (IGNORE_CHILDREN_SELECTORS && elem.webkitMatchesSelector(IGNORE_CHILDREN_SELECTORS));
    }

    /**
     * get computed styles of element, and hash them
     * @param {HTMLElement} elem
     * @returns {string}
     */
    function getStyles(elem){
        var ret = [];
        var filters = STYLE_FILTERS.slice(0);
        if(igonreChildren(elem)){
            filters.width = true;
            filters.height = true;
        }
        var styles = elem.ownerDocument.defaultView.getComputedStyle( elem, null );
        var display = styles.getPropertyValue('display');
        var opacity = styles.getPropertyValue('opacity');
        var visibility = styles.getPropertyValue('visibility');
        if(display === 'none' || opacity === '0' || visibility === 'hidden'){
            return false;
        } else {
            var position = styles.getPropertyValue('position');
            if(position !== 'static'){
                filters.push('top', 'right', 'bottom', 'left');
            }
            filters.forEach(function(key){
                ret.push(styles.getPropertyValue(key));
            });
        }
        return md5(ret.join('~'));
    }

    /**
     * get element bounding rect
     * @param {HTMLElement} elem
     * @returns [x, y, width, height]
     */
    function getRect(elem){
        var rect = elem.getBoundingClientRect();
        var doc = elem.ownerDocument;
        var win = doc.defaultView;
        var html = doc.documentElement;
        var x = Math.floor(rect.left + win.pageXOffset - html.clientLeft);
        var y = Math.floor(rect.top + win.pageYOffset - html.clientTop);
        var w = Math.floor(rect.width);
        var h = Math.floor(rect.height);
        return [x, y, w, h];
    }

    /**
     * get attributes of element
     * @param {HTMLElement} elem
     * @returns {Object|boolean}
     */
    function getAttr(elem){
        var ret = {};
        var filters = ATTR_FILTERS.slice(0);
        var hasAttr = false;
        if(elem.tagName.toLowerCase() === 'input'){
            filters.push('type');
        }
        filters.forEach(function(key){
            var attr = elem.getAttribute(key);
            if(attr !== null){
                hasAttr = true;
                ret[key] = attr;
            }
        });
        return hasAttr ? ret : false;
    }

    /**
     * filter elements
     * @param {HTMLElement} elem
     * @param {HTMLElement} parent
     * @returns {boolean}
     */
    function filter(elem, parent){
        var ret = true;
        switch (elem.nodeType){
            case 1:
                if(EXCLUDE_SELECTORS){
                    ret = ret && !elem.webkitMatchesSelector(EXCLUDE_SELECTORS);
                }
                if(INCLUDE_SELECTORS){
                    ret = ret && elem.webkitMatchesSelector(INCLUDE_SELECTORS);
                }
                break;
            case 3:
                if(IGNORE_TEXT_SELECTORS){
                    ret = ret && !parent.webkitMatchesSelector(IGNORE_TEXT_SELECTORS);
                }
                break;
            default:
                ret = false;
                break;
        }
        return ret;
    }

    /**
     * walk dom tree
     * @param {HTMLElement} elem
     * @returns {Object}
     */
    function walk(elem){
        var node = {};
        if(elem.nodeType === 1){    // element
            node.name = elem.tagName.toLowerCase();
            if(!isInvisible(elem)){
                node.rect = getRect(elem);
                var attr = getAttr(elem);
                if(attr){
                    node.attr = attr;
                }
                node.style = getStyles(elem);
                node.child = [];
                if(igonreChildren(elem)){ // ignore children
                    if(!(IGNORE_TEXT_SELECTORS && elem.webkitMatchesSelector(IGNORE_TEXT_SELECTORS))){
                        // not ignore text
                        node.child.push({
                            name: '#',
                            text: md5(elem.innerText.replace(/\s+/g, ' '))
                        });
                    }
                } else {
                    for(var i = 0, len = elem.childNodes.length; i < len; i++){
                        var child = elem.childNodes[i];
                        if(filter(child, elem)){    // recursion
                            var vdom = arguments.callee(child);  //
                            if(typeof vdom !== 'undefined' && vdom.style !== false){
                                node.child.push(vdom);
                            }
                        }
                    }
                }
                return node;
            }
        } else if(elem.nodeType === 3) {    // text node
            var text = elem.nodeValue.trim();
            if(text){
                node.name = '#';
                node.text = md5(text);
                return node;
            }
        }
    }

    if(data.removeSelectors && data.removeSelectors.length){
        data.removeSelectors.forEach(function(selector){
            var elems = document.querySelectorAll(selector);
            for(var i = 0, len = elems.length; i < len; i++){
                var elem = elems[i];
                elem.parentNode.removeChild(elem);
                elem = null;
            }
        });
    }
    return walk(document.querySelector(ROOT));

};

index.js
var child_process = require('child_process');
var spawn = child_process.spawn;
var fs = require('fs');
var path = require('path');
var Url = require('url');
var util = require("util");
var DEFAULT_DATA_DIRNAME = process.cwd();
var PHANTOMJS_SCRIPT_DIR = path.join(__dirname, 'phantomjs');
var PHANTOMJS_SCRIPT_FILE = path.join(PHANTOMJS_SCRIPT_DIR, 'index.js');
var _ = require('./util.js');
var _exists = fs.existsSync || path.existsSync;

/**
 * mkdir -p
 * @param {String} path
 * @param {Number} mode
 */
function mkdirp(path, mode){
    if (typeof mode === 'undefined') {
        //511 === 0777
        mode = 511 & (~process.umask());
    }
    if(_exists(path)) return;
    path.replace(/\\/g, '/').split('/').reduce(function(prev, next) {
        if(prev && !_exists(prev)) {
            fs.mkdirSync(prev, mode);
        }
        return prev + '/' + next;
    });
    if(!_exists(path)) {
        fs.mkdirSync(path, mode);
    }
}

/**
 * base64 encode
 * @param {String|Buffer} data
 * @returns {String}
 */
function base64(data){
    if(data instanceof Buffer){
        //do nothing for quickly determining.
    } else if(data instanceof Array){
        data = new Buffer(data);
    } else {
        //convert to string.
        data = new Buffer(String(data || ''));
    }
    return data.toString('base64');
}

/**
 * merge settings
 * @param {Object} settings
 * @returns {*}
 */
function mergeSettings(settings){
    var defaultSettings = {
        // remove phantomejs application cache before capture
        // @see https://github.com/fouber/page-monitor/issues/3
        cleanApplicationCache: false,
        // phantom cli options
        // @see http://phantomjs.org/api/command-line.html
        cli: {
            '--max-disk-cache-size' : '0',
            '--disk-cache' : 'false'
        },
        // webpage settings
        // @see http://phantomjs.org/api/webpage/
        page: {
            viewportSize: {
                width: 320,
                height: 568
            },
            settings: {
                resourceTimeout: 20000,
                userAgent: 'Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X; en-us) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53'
            }
        },
        walk: {
            invisibleElements : [
                'applet', 'area', 'audio', 'base', 'basefont',
                'bdi', 'bdo', 'big', 'br', 'center', 'colgroup',
                'datalist', 'form', 'frameset', 'head', 'link',
                'map', 'meta', 'noframes', 'noscript', 'optgroup',
                'option', 'param', 'rp', 'rt', 'ruby', 'script',
                'source', 'style', 'title', 'track', 'xmp'
            ],
            ignoreChildrenElements: [
                'img', 'canvas', 'input', 'textarea', 'audio',
                'video', 'hr', 'embed', 'object', 'progress',
                'select', 'table'
            ],
            styleFilters: [
                'margin-left', 'margin-top', 'margin-right', 'margin-bottom',
                'border-left-color', 'border-left-style', 'border-left-width',
                'border-top-color', 'border-top-style', 'border-top-width',
                'border-right-color', 'border-right-style', 'border-right-width',
                'border-bottom-color', 'border-bottom-style', 'border-bottom-width',
                'border-top-left-radius', 'border-top-right-radius',
                'border-bottom-left-radius', 'border-bottom-right-radius',
                'padding-left', 'padding-top', 'padding-right', 'padding-bottom',
                'background-color', 'background-image', 'background-repeat',
                'background-size', 'background-position',
                'list-style-image', 'list-style-position', 'list-style-type',
                'outline-color', 'outline-style', 'outline-width',
                'font-size', 'font-family', 'font-weight', 'font-style', 'line-height',
                'box-shadow', 'clear', 'color', 'display', 'float', 'opacity', 'text-align',
                'text-decoration', 'text-indent', 'text-shadow', 'vertical-align', 'visibility',
                'position'
            ],
            // attributes to mark an element
            attributeFilters: [ 'id', 'class' ],
            includeSelectors: [],
            excludeSelectors: [],
            removeSelectors: [],    // remove elements before walk
            ignoreTextSelectors: [],
            ignoreChildrenSelectors: [],
            root: 'body'
        },
        diff: {
            // LCS diff priority, `head` or `tail`
            priority: 'head',
            // highlight mask styles
            highlight: {
                add: {
                    title: '新增(Added)',
                    backgroundColor: 'rgba(127, 255, 127, 0.3)',
                    borderColor: '#090',
                    color: '#060',
                    textShadow: '0 1px 1px rgba(0, 0, 0, 0.3)'
                },
                remove: {
                    title: '删除(Removed)',
                    backgroundColor: 'rgba(0, 0, 0, 0.5)',
                    borderColor: '#999',
                    color: '#fff'
                },
                style: {
                    title: '样式(Style)',
                    backgroundColor: 'rgba(255, 0, 0, 0.3)',
                    borderColor: '#f00',
                    color: '#f00'
                },
                text: {
                    title: '文本(Text)',
                    backgroundColor: 'rgba(255, 255, 0, 0.3)',
                    borderColor: '#f90',
                    color: '#c30'
                }
            }
        },
        events: {
            init: function(token){
                /*
                    do something before page init,
                    @see http://phantomjs.org/api/webpage/handler/on-initialized.html
                */
            },
            beforeWalk: function(token){
                /*
                    do something before walk dom tree,
                    retrun a number to delay screenshot
                 */
            }
        },
        render: {
            format: 'jpeg',     // @see http://phantomjs.org/api/webpage/method/render.html
            quality: 80,        // @see http://phantomjs.org/api/webpage/method/render.html
            ext: 'jpg',         // the same as format, if not specified
            delay: 1000,        // delay(ms) before screenshot.
            timeout: 60 * 1000  // render timeout, max waiting time
        },
        path: {
            root: DEFAULT_DATA_DIRNAME, // data and screenshot save path root

            // save path format, it can be a string
            // like this: '{hostname}/{port}/{pathname}/{query}{hash}'
            format: function(url, opt){
                return [
                    opt.hostname, (opt.port ? '-' + opt.port : ''), '/',
                    base64(opt.path + (opt.hash || '')).replace(/\//g, '.')
                ].join('');
            }
        }
    };

    // special handling of events
    if(settings && settings.events){
        _.map(settings.events, function(key, value){
            if(typeof value === 'function'){
                value = value.toString().replace(/^(function\s+)anonymous(?=\()/, '$1');
                settings.events[key] = value;
            }
        });
    }
    return _.merge(defaultSettings, settings || {});
}

/**
 *
 * @param {String} path
 * @returns {String}
 */
function escapePath(path){
    if(path === '/'){
        return '-';
    } else {
        return path.replace(/^\//, '').replace(/^\.|[\\\/:*?"<>|]/g, '-');
    }
}

/**
 * path format
 * @param {String|Function} pattern
 * @param {String} url
 * @param {Object} opt
 * @returns {String}
 */
function format(pattern, url, opt){
    switch (typeof pattern){
        case 'function':
            return pattern(url, opt);
        case 'string':
            var pth = [];
            String(pattern).split('/').forEach(function(item){
                pth.push(item.replace(/\{(\w+)\}/g, function(m, $1){
                    return escapePath((opt[$1] || ''));
                }));
            });
            return pth.join('/');
        default :
            throw new Error('unsupport format');
    }
}

var LOG_VALUE_MAP ={};
var logTypes = (function(){
    var types = [];
    _.map(_.log, function(key, value){
        LOG_VALUE_MAP[value] = key.toLowerCase();
        types.push(_.escapeReg(value));
    });
    return types.join('|');
})();
var LOG_SPLIT_REG = new RegExp('(?:^|[\r\n]+)(?=' + logTypes + ')');
var LOG_TYPE_REG = new RegExp('^(' + logTypes + ')');

/**
 * why not events.EventEmitter?
 * because it can NOT emit an 'error' event,
 * but i need, fuck off.
 * @constructor
 */
var EventEmitter = function(){
    this._listeners = {};
};

/**
 * add event listener
 * @param {string} type
 * @param {Function} callback
 */
EventEmitter.prototype.on = function(type, callback){
    if(!this._listeners.hasOwnProperty(type)){
        this._listeners[type] = [];
    }
    this._listeners[type].push(callback);
};

/**
 * remove event listener
 * @param {string} type
 * @param {Function} callback
 */
EventEmitter.prototype.off = function(type, callback){
    if(this._listeners.hasOwnProperty(type)){
        var listeners = [];
        for(var i = 0, len = this._listeners[type].length; i < len; i++){
            var listener = this._listeners[type][i];
            if(listener !== callback){
                listeners.push(listener);
            }
        }
        this._listeners[type] = listeners;
    }
};

/**
 * dispatch event
 * @param {string} type
 */
EventEmitter.prototype.emit = function(type){
    if(this._listeners.hasOwnProperty(type)){
        var args = [].splice.call(arguments, 1);
        var self = this;
        this._listeners[type].forEach(function(callback){
            callback.apply(self, args);
        });
    }
};

/**
 * Monitor Class Constructor
 * @param {String} url
 * @param {Object} options
 * @constructor
 */
var Monitor = function(url, options){
    EventEmitter.call(this);
    options = mergeSettings(options);
    this.url = options.url = url;
    this.running = false;
    options.path.dir = path.join(
        options.path.root || DEFAULT_DATA_DIRNAME,
        format(options.path.format, url, Url.parse(url))
    );
    if(!fs.existsSync(options.path.dir)){
        mkdirp(options.path.dir);
    }
    this.options = options;
    this._initLog();
};

// inherit from EventEmitter
util.inherits(Monitor, EventEmitter);

/**
 * init log
 * @private
 */
Monitor.prototype._initLog = function(){
    var log = this.log = {};
    _.map(_.log, function(key){
        log[key.toLowerCase()] = [];
    });
};

/**
 * capture webpage and diff
 * @param {Function} callback
 * @param {Boolean} noDiff
 * @returns {*}
 */
Monitor.prototype.capture = function(callback, noDiff){
    if(this.running) return;
    this.running = true;
    var self = this;
    var type = _.mode.CAPTURE;
    if(!noDiff){
        type |= _.mode.DIFF;
    }
    this._initLog();
    return this._phantom(
        [
            PHANTOMJS_SCRIPT_FILE,
            type,
            this.url,
            JSON.stringify(this.options)
        ],
        function(code, log){
            // TODO with code
            self.running = false;
            callback.call(self, code, log);
        }
    );
};

/**
 * diff with two times
 * @param {Number|String|Date} left
 * @param {Number|String|Date} right
 * @param {Function} callback
 * @returns {*}
 */
Monitor.prototype.diff = function(left, right, callback){
    if(this.running) return;
    this.running = true;
    var self = this;
    var type = _.mode.DIFF;
    this._initLog();
    if(_.is(left, 'Date')){
        left = left.getDate();
    }
    if(_.is(right, 'Date')){
        right = right.getDate();
    }
    return this._phantom(
        [
            PHANTOMJS_SCRIPT_FILE,
            type, left, right,
            JSON.stringify(this.options)
        ],
        function(code, log){
            // TODO with code
            self.running = false;
            callback.call(self, code, log);
        }
    );
};

/**
 * spawn phantom
 * @param {Array} args
 * @param {Function} callback
 * @returns {*}
 * @private
 */
Monitor.prototype._phantom = function(args, callback){
    var arr = [];
    _.map(this.options.cli, function(key, value){
        arr.push(key + '=' + value);
    });
    this.emit('debug', 'cli arguments: ' + JSON.stringify(arr));
    arr = arr.concat(args);
    var proc = spawn('phantomjs', arr);
    proc.stdout.on('data', this._parseLog.bind(this));
    proc.stderr.on('data', this._parseLog.bind(this));
    proc.on('exit', function(code){
        callback(code);
    });
    return proc;
};

/**
 * parse log from phantom
 * @param {String} msg
 * @private
 */
Monitor.prototype._parseLog = function(msg){
    var self = this;
    String(msg || '').split(LOG_SPLIT_REG).forEach(function(item){
        item = item.trim();
        if(item){
            var type = 'debug';
            item = item.replace(LOG_TYPE_REG, function(m, $1){
                type = LOG_VALUE_MAP[$1] || type;
                return '';
            });
            self.emit(type, item);
            if(self.log.hasOwnProperty(type)){
                self.log[type].push(item);
            }
        }
    });
};

module.exports = Monitor;